File: Python/combination-sum-ii.py
class Solution:
    # @param candidates, a list of integers
    # @param target, integer
    # @return a list of lists of integers
    def combinationSum2(self, candidates, target):
        ans = []
        n = len(candidates)
        candidates.sort()
        
        def DFS(candidates, target, start, partial_sol):
            if target==0 and partial_sol not in ans:
                ans.append(partial_sol)
            for i in xrange(start, n):
                if target<candidates[i]: return
                if i==start or candidates[i-1]!=candidates[i]:
                    DFS(candidates, target-candidates[i], i+1, partial_sol+[candidates[i]])
        
        DFS(candidates, target, 0, [])
        return ans

File: C++/combination-sum-ii.cpp
class Solution {
private:
    void DFS(std::vector<int> &candidates, int start, int target, std::vector<int> &partial, std::vector<std::vector<int>> &ans) {
        if(target==0) {
            ans.push_back(partial);
            return;
        }
        for(int i=start; i<candidates.size() && target>=candidates[i]; ++i) {
            if(i==start || candidates[i-1]!=candidates[i]) {
                partial.push_back(candidates[i]);
                DFS(candidates, i+1, target-candidates[i], partial, ans);
                partial.pop_back();
            }
        }
    }
public:
    std::vector<std::vector<int>> combinationSum2(std::vector<int> &candidates, int target) {
        std::sort(candidates.begin(), candidates.end());
        std::vector<int> partial;
        std::vector<std::vector<int>> ans;
        DFS(candidates, 0, target, partial, ans);
        return ans;
    }
};
